### DISCUSSION TITLE
CMake support

### DISCUSSION THREAD

**Discussion by samuelpmish** [CEED/libCEED#1524] [General]
Hi, is there any interest in supporting a CMake build?

I've got a couple of mfem-adjacent projects and CMake enables automated fetching/configuring/building dependencies, which makes things easy for users. However, when some of those dependencies are makefile-only, then it requires some additional manual intervention from the user, which is less pleasant.

I'm happy to help set things up if there's interest.

**Comment by jrwrigh** [comment]
Would this effectively just be a wrapper around the Makefile? Like, just take the CMake-generated flags and pass them to make? Or would we effectively need to support two different build systems?

Side note: it'd probably be good to make sure that even in the CMake route, `pkgconfig` files are still created and handled correctly. SmartRedis [I know](https://github.com/CrayLabs/SmartRedis/issues/333) has this functionality. Not sure if there's a better way to do it though. Perhaps we can just copy the Makefile-generated `ceed.pc` file to wherever it needs to be.

**Reply by samuelpmish** [reply]
It wouldn't wrap the existing makefile-- the point of CMake is to specify the sources/options for a project and then it generates either a makefile, build.ninja, or IDE project from that. It's possible to have it coexist alongside the existing makefile, although that does potentially run the risk of the two going out of sync over time.

re: `pkgconfig` files, those can also be generated with CMake and copied wherever is appropriate as part of the `install` process.

Concretely, some of the benefits of a CMake build are:
1. cross platform support (although this likely doesn't matter for libCEED)
2. other CMake projects can automatically download/build/link against libCEED in their project by adding a couple lines to their CMakeLists.txt
3. less manual configuration required by the user, for example
    - CMake can find where the CUDA toolkit is installed, so users don't need to specify custom directories
    - CMake can reflect on the OS, hardware and compiler to find out which flags to pass
4. a way to wrap subprojects and propagate options to them from a parent project

As a library consumer, I'm (selfishly) most interested in points 2, 3 and 4!

**Reply by jeremylt** [reply]
I don't think we want to replace our Make build system with CMake. I'm definitely open to supporting CMake (though I don't use it personally). I would want to make (heh) sure that such a thing is tested, perhaps via a GitHub action.

**Reply by nbeams** [reply]
~The projects I know of that offer both Makefile and CMake do effectively support two different build systems (including MFEM itself).~ Edit: this was already answered while I was typing :) 

@samuelpmish, in the meantime, would the CMake modules available in MFEM itself (`FindlibCEED.cmake`, which uses `MfemCMakeUtilities.cmake`) help for your projects? When using the CMake build of MFEM, one only needs to specify `MFEM_USE_CEED` and `CEED_DIR` during configuration, and MFEM's CMake setup takes care of the rest.

**Reply by samuelpmish** [reply]
@nbeams I'm aware that mfem wrote a tool to look for libCEED, although if I have to specify exactly where libCEED is for it to work, that sort of defeats the purpose! Also, it doesn't help with actually building libCEED with the appropriate flags for one's system, or linking it into the parent project.

Maybe it'd be helpful for me to make a PR so there's an implementation to try out.

**Reply by nbeams** [reply]
I'm not sure I'd agree that it "defeats the purpose" from a user standpoint, as including a package with only minor changes to the CMake configuration is pretty easy. Depending on where things are installed and whether there are multiple versions of a particular dependency, users might specify the location of other CMake-built dependencies of MFEM (or need to add to CMake search path), as well. And I don't see how it wouldn't help with linking it in to the parent project, either as part of MFEM's dependencies, or (as I was actually thinking) if the parent project were the one doing the searching, using the MFEM modules as a base just to avoid reinventing the wheel.

Mostly I just didn't want this to be a blocking issue for your other projects, but if you want to work on the CMake build yourself, then it doesn't matter.
